### PR TITLE
simple scale buttons for plots

### LINKED ISSUE #202: Transform Table Plot Sizes
with the newly added scree plot (#182). The Transform Table looks like this on a regular Full HD screen: 

image

possible ways to deal with that:
 * allow to resize the plots in a granular way (e.g. small, medium, large). a  CSS grid layout could be used for layouting
 * allow to en/disable individual plots
 * auto adapt the plots and plot sizes to the available screen (disadvantage: unbalanced aspect ratios)

### PR DESCRIPTION
**Pull request by sgratzl**
closes #202 by adding scale buttons to plots and option to disable the plot

image

image

### CODE REVIEW

**Comment by subdavis** [comment]
Neat, I think something like this would be handly.

I had a similar idea, but it involved swapping between `vis-tile` and `vis-tile-large` such that there were only 2 sizes of plot: default and full width.  The nice part about that would be, with the roomier `vis-tile-large`, the menu controls could automatically jump into a left nav pane.

Regardless of implementation, I think providing more than those two options may be overkill, but I would like to hear your opinion.

**Comment by subdavis** [comment]
You could imagine something a bit like this, but with the plot the appropriate height and width.

Screenshot from 2019-09-19 10-40-43

**Comment by sgratzl** [comment]
@subdavis can you please your comments to #202. The issue is a better place to discuss how the issue should be implemented

**Comment by sgratzl** [comment]
grid compacting (works best with more plots)

image

**Comment by sgratzl** [comment]
image

image


when one disables and enables the plot, the plot won't be rerended. @waxlamp do you have an idea why?

**Comment by subdavis** [comment]
@sgratzl it's because of a bug @waxlamp and I talked about in #195 where the plot doesn't render at mount.

I think it should be fixed on master.  Is it possible to rebase and try again?

**Comment by subdavis** [comment]
You could also sidestep the issue by using v-show instead of v-if for now and fix it later.

**Comment by sgratzl** [comment]
extracted bug to https://github.com/girder/viime/issues/244

File: web/src/components/LayoutGrid.vue [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+.gridlayout(:style="gridStyle")
+  slot
+</template>
+
+<style scoped lang="scss">
+.gridlayout {
+  display: grid;
+  grid-gap: 0.5em;
+}
+</style>
```
Comment (subdavis) [inline web/src/components/LayoutGrid.vue]: +1  for making this a separate component.

File: web/src/components/Transform.vue [hunk 2]
```diff
@@ -27,6 +29,13 @@ export default {
       normalize_methods,
       transform_methods,
       scaling_methods,
+      visiblePlots: {
+        score: true,
+        scree: true,
+        loadings: true,
+      },
+      cellSize: 300,
+      plotHeader: 48 + 5,
```
Comment (subdavis) [inline web/src/components/Transform.vue]: Could you add a comment about where 48 and 5 came from?  +1 for separating them out an not just writing `53`.

File: web/src/components/vis/LoadingsPlotTile.vue [hunk 3]
```diff
@@ -51,13 +51,14 @@ export default {
...
 <template lang="pug">
 vis-tile(v-if="plot", title="PCA Loadings Plot", :loading="plot.loading")
-  loadings-plot(
-      :width="width",
-      :height="height",
-      :points="plot.data",
-      :pc-x="pcX",
-      :pc-y="pcY",
-      :show-crosshairs="showCrosshairs")
+  template(#default="wrapper")
```
Comment (subdavis) [inline web/src/components/vis/LoadingsPlotTile.vue]: Is this change relevant to this PR?   It's fine if it isn't, just curious.

File: web/src/components/vis/VisTile.vue [hunk 4]
```diff
@@ -11,28 +11,61 @@ export default {
...
+    };
+  },
+
   computed: {
-    hasControls() {
-      return !!this.$slots.controls;
+    smallEnough() {
+      return this.scale <= 0.5;
+    },
+    largeEnough() {
```
Comment (subdavis) [inline web/src/components/vis/VisTile.vue]: Prefer better names.  perhaps `isMaxScale` and `isMinScale`.

File: web/src/components/vis/VisTile.vue [hunk 5]
```diff
@@ -11,28 +11,61 @@ export default {
...
+      return this.scale >= 4;
+    },
+    scaleClass() {
+      return `scale${Math.round(this.scale * 100)}`;
```
Comment (subdavis) [inline web/src/components/vis/VisTile.vue]: `* 100` seems unnecessary.  Why not just have your min and max be 50 and 400 respectively?

Actually, why was multiplication used here? if the point is to have N different classes which apply N different grid sizes, why not just use addition?  `scale += 1` and `scale -= 1` would achieve the same thing without the risk of floating point issues.  You're already stuck on an exponential scale, so it's not like you could add scale=500 if you wanted something in the middle anyway.

File: web/src/components/vis/VisTile.vue [hunk 6]
```diff
@@ -11,28 +11,61 @@ export default {
...
+  },
+
   computed: {
-    hasControls() {
-      return !!this.$slots.controls;
+    smallEnough() {
+      return this.scale <= 0.5;
+    },
+    largeEnough() {
+      return this.scale >= 4;
```
Comment (subdavis) [inline web/src/components/vis/VisTile.vue]: Perhaps refactor `minScale` and `maxScale` to constants at the top of the script.

File: web/src/components/vis/VisTile.vue [hunk 7]
```diff
@@ -11,28 +11,61 @@ export default {
...
+}
+.scale100 {
+  grid-column: span 2;
+  grid-row: span 2;
+}
+.scale200 {
+  grid-column: span 4;
+  grid-row: span 4;
+}
+.scale400 {
```
Comment (subdavis) [inline web/src/components/vis/VisTile.vue]: prefer `scale1, scale2, scale3` etc.  The actual numbers don't seem to matter, so you could use grid layouts that aren't powers of 2.

File: web/src/components/vis/LoadingsPlotTile.vue (hunk 3 above)
Comment (sgratzl) [reply web/src/components/vis/LoadingsPlotTile.vue]: I'm accessing the `wrapper.scale` one line below and I don't know another way to assign a scoped prop

File: web/src/components/vis/VisTile.vue (hunk 4 above)
Comment (sgratzl) [reply web/src/components/vis/VisTile.vue]: those methods just exists because of limitation of pug

File: web/src/components/vis/VisTile.vue (hunk 7 above)
Comment (sgratzl) [reply web/src/components/vis/VisTile.vue]: `span 4` refers to that four cells should be used. `scale200` mean a scaling factor of  200 percent. Thus, you can read it as: When the scaling factor is `200`% the cell should span over four cells in the grid layout

Comment (subdavis) [reply web/src/components/vis/VisTile.vue]: I suppose my point is that the math here, while technically correct, isn't necessary.  You've already hard-coded the result of the equation `cell width * 200%` within the CSS class.

All you really have is a finite set of constant classes you can apply, and you iterate through them by generating their string keys mathematically. 

I think you could reduce the complexity by having something like below where you just have buttons that increment a counter up and down.  A bonus here would be that your min and max go away, and you iterate from 0 to `sizeClasses.length - 1`

```
<script>
  ...
  data() { return { sizeClasses = [ 'class1', 'class2', 'class3', ...] }},
<template>
   ...
   <element :class="sizeClasses[index]"
</template>
```

File: web/src/components/vis/VisTile.vue (hunk 6 above)
Comment (subdavis) [reply web/src/components/vis/VisTile.vue]: This discussion has moved to the thread below.

File: web/src/components/vis/VisTile.vue (hunk 5 above)
Comment (subdavis) [reply web/src/components/vis/VisTile.vue]: Move this conversation to the thread below.

File: web/src/components/vis/VisTile.vue (hunk 7 above)
Comment (subdavis) [reply web/src/components/vis/VisTile.vue]: Another consideration is that the math ties us unnecessarily to this exponential scale.  We might find that a Fibonacci scale is more appropriate, and if the class name generation weren't mathematical, all you have to do is change a few CSS properties.

Comment (waxlamp) [reply web/src/components/vis/VisTile.vue]: +1 for simplifying the logic here.

File: web/src/components/vis/VisTile.vue [hunk 8]
```diff
@@ -11,28 +11,61 @@ export default {
...
   },
 
+  data() {
+    return {
+      scale: 1,
+    };
+  },
+
   computed: {
-    hasControls() {
```
Comment (waxlamp) [inline web/src/components/vis/VisTile.vue]: It doesn't seem like this should have been deleted--it's meant to control whether the menu icon appears in the toolbar.

Comment (sgratzl) [reply web/src/components/vis/VisTile.vue]: no it controlled whether the overall toolbar on top was rendered. The menu icon is rendered in the Plot tiles directly.

**Comment by sgratzl** [comment]
changed version:

image

**Comment by sgratzl** [comment]
feedback is integrated

File: web/src/components/vis/VisTile.vue [hunk 9]
```diff
@@ -11,28 +13,91 @@ export default {
...
     v-toolbar.primary.darken-3.top-rounded(dark, flat, dense)
       v-toolbar-title {{ title }}
       v-spacer
-      v-toolbar-items(v-if="hasControls")
+      v-toolbar-items
+        v-btn(@click="scaleIndex--", :disabled="scaleIndex === 0", icon)
+          v-icon {{ $vuetify.icons.magnifyMinus }}
+        v-select.scaleFactor(:value="scaleIndex", @change="scaleIndex = parseInt($event, 10)",
+            :items="scaleOptions", hide-details, item-text="text", item-value="value")
+        v-btn(@click="scaleIndex++", :disabled="scaleIndex === scales.length - 1", icon)
```
Comment (subdavis) [inline web/src/components/vis/VisTile.vue]: Better not to rely on the Vue event loop to protect you from invalid states.  If the page or the machine is under heavy load, `:disabled` may not compute fast enough. 

Almost every project I've got right now has [this](https://stackoverflow.com/questions/11409895/whats-the-most-elegant-way-to-cap-a-number-to-a-segment) in the utilities.

Recommend:

```
script
 import { clamp } from 'utils';
  methods
    setScaleIndex(value) { this.scaleIndex = clamp(value, min, max); },
...
template
  v-btn(@click="setScaleIndex(scaleIndex +/- 1)", ...)
```

Comment (subdavis) [reply web/src/components/vis/VisTile.vue]: To be clear, I don't use the Number prototype, I just take the number as the first argument.

File: web/src/components/Transform.vue (hunk 2 above)
Comment (sgratzl) [reply web/src/components/Transform.vue]: can be removed when https://github.com/girder/viime/issues/256 is implemented

File: web/src/components/vis/VisTile.vue (hunk 9 above)
Comment (sgratzl) [reply web/src/components/vis/VisTile.vue]: done
